packages/enhanced/src/lib/container/HoistContainerReferencesPlugin.ts

```diff
@@ -4,27 +4,143 @@ import type {
   Chunk,
   WebpackPluginInstance,
 } from 'webpack';
+import { normalizeWebpackPath } from '@module-federation/sdk/normalize-webpack-path';
 import ContainerEntryModule from './ContainerEntryModule';
 
+const runtime = require(
+  normalizeWebpackPath('webpack/lib/util/runtime'),
+) as typeof import('webpack/lib/util/runtime');
+
 /**
  * This class is used to hoist container references in the code.
  * @constructor
  */
-export class HoistContainerReferences implements WebpackPluginInstance {
+export class HoistContainerReferencesPlugin implements WebpackPluginInstance {
+  integrateChunks(
+    chunkA: Chunk,
+    chunkB: Chunk,
+    compilation: Compilation,
+  ): void {
+    const { chunkGraph, compiler } = compilation;
+    // do not sort chunk by smallest name, this will cause non-deterministic chunk integration
+
+    // if (chunkA.name && chunkB.name) {
+    //   if (
+    //     chunkGraph.getNumberOfEntryModules(chunkA) > 0 ===
+    //     chunkGraph.getNumberOfEntryModules(chunkB) > 0
+    //   ) {
+    //     // When both chunks have entry modules or none have one, use
+    //     // shortest name
+    //     if (chunkA.name.length !== chunkB.name.length) {
+    //       chunkA.name =
+    //         chunkA.name.length < chunkB.name.length ? chunkA.name : chunkB.name;
+    //     } else {
+    //       chunkA.name = chunkA.name < chunkB.name ? chunkA.name : chunkB.name;
+    //     }
+    //   } else if (chunkGraph.getNumberOfEntryModules(chunkB) > 0) {
+    //     // Pick the name of the chunk with the entry module
+    //     chunkA.name = chunkB.name;
+    //   }
+    // } else if (chunkB.name) {
+    //   chunkA.name = chunkB.name;
+    // }
+    //
+    // // Merge id name hints
+    for (const hint of chunkB.idNameHints) {
+      chunkA.idNameHints.add(hint);
+    }
+
+    // Merge runtime
+    //@ts-ignore
+    chunkA.runtime = runtime.mergeRuntime(chunkA.runtime, chunkB.runtime);
+
+    // getChunkModules is used here to create a clone, because disconnectChunkAndModule modifies
+    for (const module of chunkGraph.getChunkModules(chunkB)) {
+      // if (chunkB.name !== 'federation-runtime') {
+      chunkGraph.disconnectChunkAndModule(chunkB, module);
+      // }
+      chunkGraph.connectChunkAndModule(chunkA, module);
+    }
+
+    for (const [module, chunkGroup] of Array.from(
+      chunkGraph.getChunkEntryModulesWithChunkGroupIterable(chunkB),
+    )) {
+      // chunkGraph.disconnectChunkAndEntryModule(chunkB, module);
+      chunkGraph.connectChunkAndModule(chunkA, module);
+      // chunkGraph.connectChunkAndEntryModule(chunkA, module,chunkGroup);
+    }
+
+    for (const chunkGroup of chunkB.groupsIterable) {
+      chunkGroup.replaceChunk(chunkB, chunkA);
+      chunkA.addGroup(chunkGroup);
+      chunkB.removeGroup(chunkGroup);
+    }
+    compiler.webpack.ChunkGraph.clearChunkGraphForChunk(chunkB);
```

Comment:
if the project has multi entry : main,main2 , can main2 entry integrate successfully while the chunkB has been cleared in main entry ?

Reply:
I know this works for multiple runtime chunks - but let me add a test for multiple entrypoints. 
I imagine the same will have when each entrypoint includes runtime chunk and not single runtime or split runtime chunk.
`multiple-runtime-chunk`
<img width="1381" alt="image" src="https://github.com/module-federation/universe/assets/25274700/cb8fc32e-6637-4840-a095-9ac527adf3c4">
<img width="1381" alt="image" src="https://github.com/module-federation/universe/assets/25274700/3ca03223-de7a-4f63-a5ef-9545f2fb69c1">


I added for multi entry, youre right - there are missing modules but the two entry modules are there, but bundler runtime is not in the second entry chunk.

Reply:
Added test case and patched plugin accordingly. I will not disconnect the module from chunks, then i can loop over chunks again and copy them multiple times - then at end of the compile i just delete the chunk and that cleans it up anyways.

Reply:
So the issue is not clearChunkGraph, but this:

```js
for (const module of chunkGraph.getChunkModules(chunkB)) {
      // dont disconnect as module may need to be copied into multiple chunks
      chunkGraph.disconnectChunkAndModule(chunkB, module);
      chunkGraph.connectChunkAndModule(chunkA, module);
    }
    ```
    
    Clear chunkGraph ensures the chunk is disconnected from the graph.
    We want to detach the chunk from the graph, and use its modus as a cache/reference to copy into runtime chunks. Then we delete the chunk. 
    
    If we dont clear chunk graph then other chunks my reference a chunk that no longer exists